p2p/protocol/head/head.go

```diff
@@ -0,0 +1,107 @@
+package head
+
+import (
+	"context"
+	"encoding/json"
+	"net"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	gostream "github.com/libp2p/go-libp2p-gostream"
+)
+
+var log = logging.Logger("go-legs-head")
+
+func deriveProtocolID(topic string) protocol.ID {
+	return protocol.ID("/legs/head/" + topic + "/0.0.1")
+}
+
+func (p *Publisher) Serve(ctx context.Context, host host.Host, topic string) error {
+	l, err := gostream.Listen(host, deriveProtocolID(topic))
+	if err != nil {
+		return err
+	}
+
+	p.server = &http.Server{Handler: http.Handler(p)}
+	errChan := make(chan error)
+	go func() {
+		errChan <- p.server.Serve(l)
+	}()
```

Comment:
I think that instead of creating a go routine, you do:
```go
return p.server.Serve(l)
``` 
and let the caller decide if they want to wait in a goroutine, it will be simpler.  Since this function is always called
```go
go p.Server(...)
```
I do not see the point of creating an internal goroutine and waiting on it.  This also means that the context becomes meaningless since it is not passed into `p.server.Serve()` to cancel that call.